staking-pool-shares/src/lib.rs

```diff
@@ -0,0 +1,802 @@
+
+construct_uint! {
+    /// 256-bit unsigned integer.
+    // TODO: Revert back to 4 once wasm/wasmer bug is fixed.
+    // See https://github.com/wasmerio/wasmer/issues/1429
+    pub struct U256(8);
+}
+
+#[cfg(test)]
+mod test_utils;
+
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+/// Inner account data of a delegate.
+#[derive(BorshDeserialize, BorshSerialize, Debug, PartialEq)]
+pub struct Account {
+    /// The unstaked balance. It represents the amount the account has on this contract that
+    /// can either be staked or withdrawn.
+    pub unstaked: Balance,
+    /// The amount of "stake" shares. Every stake share corresponds to the amount of staked balance.
+    /// NOTE: The number of shares should always be less or equal than the amount of staked balance.
+    /// This means the price of stake share should always be at least `1`.
+    /// The price of stake share can be computed as `total_staked_balance` / `total_stake_shares`.
+    pub stake_shares: Balance,
+    /// The minimum epoch height when the withdrawn is allowed.
+    /// This changes after unstaking action, because the amount is still locked for 3 epochs.
+    pub unstaked_available_epoch_height: EpochHeight,
+}
+
+impl Default for Account {
+    fn default() -> Self {
+        Self {
+            unstaked: 0,
+            stake_shares: 0,
+            unstaked_available_epoch_height: 0,
+        }
+    }
+}
+
+/// The number of epochs required for the locked balance to become unlocked.
+const NUM_EPOCHS_TO_UNLOCK: EpochHeight = 3;
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct StakingContract {
+    /// The account ID of the owner who's running the staking validator node.
+    /// NOTE: This is different from the current account ID which is used as a validator account.
+    /// The owner of the staking pool can change staking public key and adjust reward fees.
+    pub owner_id: AccountId,
+    /// The public key which is used for staking action. It's the public key of the validator node
+    /// that validates on behalf of the pool.
+    pub stake_public_key: PublicKey,
+    /// The last epoch height when `ping` was called.
+    pub last_epoch_height: EpochHeight,
+    /// The last known amount of locked balance of this account.
+    pub last_locked_account_balance: Balance,
+    /// The last known amount of unlocked balance of this account.
+    pub last_account_balance: Balance,
+    /// The total amount of shares. It should be equal to the total amount of shares across all
+    /// accounts.
+    pub total_stake_shares: Balance,
+    /// The total staked balance.
+    pub total_staked_balance: Balance,
+    /// The fraction of the reward that goes to the owner of the staking pool for running the
+    /// validator node.
+    pub reward_fee_fraction: RewardFeeFraction,
+    /// Persistent map from an account ID hash to the corresponding account.
+    pub accounts: Map<AccountHash, Account>,
```

Comment:
Why do we need to use hash here instead of `AccountId`?

Reply:
To avoid slowing lookups with account collisions. Since the account ID is limited to 64 characters, the most you can slow down the lookup is by 64 nodes. So it's not critical to do this for account ID. Still, it's the best practice to hash user-input when storing internally, to avoid creating deep tries.